format/postgres/common/pg_btree/postgres/pg_btree.go

```diff
@@ -0,0 +1,284 @@
+package postgres
+
+import (
+	"github.com/wader/fq/format"
+	"github.com/wader/fq/format/postgres/common"
+	"github.com/wader/fq/format/postgres/common/pg_heap/postgres"
+	"github.com/wader/fq/pkg/decode"
+)
+
+const (
+	BTREE_MAGIC = 0x053162
+	P_NONE      = 0
+
+	/* Bits defined in btpo_flags */
+	BTP_LEAF             = 1 << 0 /* leaf page, i.e. not internal page */
+	BTP_ROOT             = 1 << 1 /* root page (has no parent) */
+	BTP_DELETED          = 1 << 2 /* page has been deleted from tree */
+	BTP_META             = 1 << 3 /* meta-page */
+	BTP_HALF_DEAD        = 1 << 4 /* empty, but still in tree */
+	BTP_SPLIT_END        = 1 << 5 /* rightmost page of split group */
+	BTP_HAS_GARBAGE      = 1 << 6 /* page has LP_DEAD tuples (deprecated) */
+	BTP_INCOMPLETE_SPLIT = 1 << 7 /* right sibling's downlink is missing */
+	BTP_HAS_FULLXID      = 1 << 8 /* contains BTDeletedPageData */
+)
+
+const (
+	INDEX_SIZE_MASK       = 0x1FFF
+	INDEX_AM_RESERVED_BIT = 0x2000 /* reserved for index-AM specific usage */
+	INDEX_VAR_MASK        = 0x4000
+	INDEX_NULL_MASK       = 0x8000
+)
+
+const (
+	IndexTupleDataSize = 8
+)
+
+// struct BTMetaPageData {
+/*    0      |     4 */ // uint32 btm_magic
+/*    4      |     4 */ // uint32 btm_version
+/*    8      |     4 */ // BlockNumber btm_root
+/*   12      |     4 */ // uint32 btm_level
+/*   16      |     4 */ // BlockNumber btm_fastroot
+/*   20      |     4 */ // uint32 btm_fastlevel
+/*   24      |     4 */ // uint32 btm_last_cleanup_num_delpages
+/* XXX  4-byte hole */
+/*   32      |     8 */ // float8 btm_last_cleanup_num_heap_tuples
+/*   40      |     1 */ // _Bool btm_allequalimage
+/* XXX  7-byte padding */
+//
+/* total size (bytes):   48 */
+
+// struct BTPageOpaqueData {
+/*    0      |     4 */ // BlockNumber btpo_prev;
+/*    4      |     4 */ // BlockNumber btpo_next;
+/*    8      |     4 */ // uint32 btpo_level;
+/*   12      |     2 */ // uint16 btpo_flags;
+/*   14      |     2 */ // BTCycleId btpo_cycleid;
+//
+/* total size (bytes):   16 */
+
+// struct IndexTupleData {
+/*    0      |     6 */ // ItemPointerData t_tid;
+/*    6      |     2 */ // unsigned short t_info;
+//
+// IndexTupleData *IndexTuple;
+/* total size (bytes):    8 */
+
+func DecodePgBTree(d *decode.D, args format.Pg_BTree_In) any {
+	btree := &BTree{
+		Args:     args,
+		PageSize: common.PageSize,
+	}
+	decodeBTreePages(btree, d)
+	return nil
+}
+
+type BTree struct {
+	Args     format.Pg_BTree_In
```

Comment:
Looks like only `Args.Page` is used, could possibly decouple a bit by accepting `page int` as arg for `DecodePgBTree`

Reply:
Maybe this can be refactor like this:
```diff
diff --git a/format/postgres/common/pg_btree/postgres/pg_btree.go b/format/postgres/common/pg_btree/postgres/pg_btree.go
index 5ae586c5..be387438 100644
--- a/format/postgres/common/pg_btree/postgres/pg_btree.go
+++ b/format/postgres/common/pg_btree/postgres/pg_btree.go
@@ -1,7 +1,6 @@
 package postgres
 
 import (
-	"github.com/wader/fq/format"
 	"github.com/wader/fq/format/postgres/common"
 	"github.com/wader/fq/format/postgres/common/pg_heap/postgres"
 	"github.com/wader/fq/pkg/decode"
@@ -65,30 +64,17 @@ const (
 // IndexTupleData *IndexTuple;
 /* total size (bytes):    8 */
 
-func DecodePgBTree(d *decode.D, args format.Pg_BTree_In) any {
-	btree := &BTree{
-		Args:     args,
-		PageSize: common.PageSize,
-	}
-	decodeBTreePages(btree, d)
-	return nil
-}
+func DecodePgBTree(d *decode.D, pageNr int) {
+	var prevPage *postgres.HeapPage
 
-type BTree struct {
-	Args     format.Pg_BTree_In
-	PageSize uint64
-	page     *postgres.HeapPage
-}
-
-func decodeBTreePages(btree *BTree, d *decode.D) {
-	for i := btree.Args.Page; ; i++ {
+	for i := pageNr; ; i++ {
 		page := &postgres.HeapPage{}
-		if btree.page != nil {
+		if prevPage != nil {
 			// use prev page
-			page.BytesPosBegin = btree.page.BytesPosEnd
+			page.BytesPosBegin = prevPage.BytesPosEnd
 		}
-		page.BytesPosEnd = int64(common.TypeAlign(btree.PageSize, uint64(page.BytesPosBegin)+1))
-		btree.page = page
+		page.BytesPosEnd = int64(common.TypeAlign(common.PageSize, uint64(page.BytesPosBegin)+1))
+		prevPage = page
 
 		pos0 := page.BytesPosBegin * 8
 		d.SeekAbs(pos0)
@@ -100,19 +86,18 @@ func decodeBTreePages(btree *BTree, d *decode.D) {
 		if i == 0 {
 			// first page contains meta information
 			d.FieldStruct("page", func(d *decode.D) {
-				decodeBTreeMetaPage(btree, d)
+				decodeBTreeMetaPage(page, d)
 			})
 			continue
 		}
 
 		d.FieldStruct("page", func(d *decode.D) {
-			decodeBTreePage(btree, d)
+			decodeBTreePage(page, d)
 		})
 	}
 }
 
-func decodeBTreeMetaPage(btree *BTree, d *decode.D) {
-	page := btree.page
+func decodeBTreeMetaPage(page *postgres.HeapPage, d *decode.D) {
 
 	d.FieldStruct("page_header", func(d *decode.D) {
 		postgres.DecodePageHeader(page, d)
@@ -122,7 +107,7 @@ func decodeBTreeMetaPage(btree *BTree, d *decode.D) {
 	})
 
 	pos0 := d.Pos()
-	pos1 := btree.page.BytesPosSpecial * 8
+	pos1 := page.BytesPosSpecial * 8
 	d.FieldRawLen("unused0", pos1-pos0)
 	d.FieldStruct("page_opaque_data", func(d *decode.D) {
 		decodeBTPageOpaqueData(d)
@@ -204,15 +189,13 @@ func decodeBTPageOpaqueData(d *decode.D) {
 	})
 }
 
-func decodeBTreePage(btree *BTree, d *decode.D) {
-	page := btree.page
-
+func decodeBTreePage(page *postgres.HeapPage, d *decode.D) {
 	d.FieldStruct("page_header", func(d *decode.D) {
 		postgres.DecodePageHeader(page, d)
 	})
 
 	pos0 := d.Pos()
-	pos1 := btree.page.BytesPosSpecial * 8
+	pos1 := page.BytesPosSpecial * 8
 	d.SeekAbs(pos1)
 	d.FieldStruct("page_opaque_data", func(d *decode.D) {
 		decodeBTPageOpaqueData(d)
@@ -227,12 +210,11 @@ func decodeBTreePage(btree *BTree, d *decode.D) {
 	postgres.DecodeItemIds(page, d)
 
 	d.FieldArray("tuples", func(d *decode.D) {
-		decodeIndexTuples(btree, d)
+		decodeIndexTuples(page, d)
 	})
 }
 
-func decodeIndexTuples(btree *BTree, d *decode.D) {
-	page := btree.page
+func decodeIndexTuples(page *postgres.HeapPage, d *decode.D) {
 
 	for i := 0; i < len(page.ItemIds); i++ {
 		id := page.ItemIds[i]
diff --git a/format/postgres/pg_btree.go b/format/postgres/pg_btree.go
index 5a53aa42..9df49ac4 100644
--- a/format/postgres/pg_btree.go
+++ b/format/postgres/pg_btree.go
@@ -32,5 +32,6 @@ func decodePgBTree(d *decode.D) any {
 	if !d.ArgAs(&pgIn) {
 		d.Fatalf("no page specified")
 	}
-	return postgres.DecodePgBTree(d, pgIn)
+	postgres.DecodePgBTree(d, pgIn.Page)
+	return nil
 }

```

Reply:
Patch applied.

Reply:
👍 maybe an older version had more things in BTree?

Reply:
No, nothing to add.